File: max-consecutive-ones-ii/max-consecutive-ones-ii.py
class Solution:
    def findMaxConsecutiveOnes(self, nums: List[int]) -> int:
        """
        sliding window
        expand: on 1 or first zero
        contract: on second zero
        """
        left = right = ans = zeroes = 0

        while right < len(nums):
            if nums[right] == 0:
                zeroes += 1

            while zeroes == 2:
                if nums[left] == 0:    
                    zeroes -= 1
                left += 1

            ans = max(ans, right - left + 1)
            right += 1

        return ans

File: max-consecutive-ones-ii/max-consecutive-ones-ii.java
class Solution {
    public int findMaxConsecutiveOnes(int[] nums) {

        int leftPointer = 0, rightPointer = 0, zeroes = 0, max=0;
        
        while (rightPointer < nums.length){
            if (nums[rightPointer] == 0){
                zeroes++;
            }
            
            if (zeroes > 1){
                if(nums[leftPointer]==0){
                    zeroes--;
                }
                leftPointer++;
            }
            
            max = Math.max(rightPointer-leftPointer+1, max);
            
            rightPointer++;
            
        }
        return max;
    }
}